### PR TITLE
Fix usage multiplication and naming

### PR DESCRIPTION
**Pull request by jonaprieto**
This PR basically ends the work started in https://github.com/anoma/juvix/pull/842 (which is now 30 commits behind. Then I suggest to close it after merging this one).

The type of usage is simply the _extended natural number system_, i.e. ℕ u {∞}. 
The previous notation ω for the unrestricted usage is misleading, and I suggest to
use a more familiar symbol for programmers `_`, which is better than my initial suggestion `∞`.

- [x] `Omega` is now `SAny`.
- [x]  `0*x = x*0 = 0`
- [x] `pred : Usage → Maybe Usage`
- [x] `predPosUsage : Usage → Usage` to fix compilation with Michelson backend. 
- [ ]  Replace `ω` by `_`

### CODE REVIEW

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/Compilation/VirtualStack.hs [hunk 1]
```diff
@@ -323,17 +323,15 @@ drop :: Int -> T lamType -> T lamType
...
-            drop (pred n) (cdr xs)
-
--- This propagates usages. This is safe, as if a var has multiple names, it thus
--- must be the same exact var
--- also unlike drop we do not drop a usage
+    case car xs of
+      (VarE x (Usage π _) _, _)
+        | Usage.isNotZero π ->
+          drop (n - 1) (updateUsage x (Usage.predPosUsage π) (cdr xs))
+      _ -> drop (n - 1) (cdr xs)
```
Comment (mariari) [inline library/Backends/Michelson/src/Juvix/Backends/Michelson/Compilation/VirtualStack.hs:330]: Keep the `pred`?

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs [hunk 2]
```diff
@@ -813,7 +813,7 @@ apply closure args remainingArgs = do
       throw @"compilationError" (Types.InternalFault "apply to non lam")
 
 -- we can only delete things at position greater than 0.
--- this is because if we were to delete 0, then (λx : ω i. x) would error
+-- this is because if we were to delete 0, then (λx : _ i. x) would error
```
Comment (mariari) [inline library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs:816]: Probably shuold find a new symbol for this, as it's not clear with this what would error

File: library/Core/src/Juvix/Core/IR/CheckTerm.hs [hunk 3]
```diff
@@ -484,7 +484,7 @@ evalTC t = do
 -- Currently subtyping consists of the following:
 --
 -- * Consistency of universe levels (@*ᵢ <: *ⱼ@ if @i ≤ j@)
--- * Usage compatibility (@(π x: A) → B <: (ω x: A) → B@ for finite @π@)
+-- * Usage compatibility (@(π x: A) → B <: (- x: A) → B@ for finite @π@)
```
Comment (mariari) [inline library/Core/src/Juvix/Core/IR/CheckTerm.hs:487]: Here as well

File: library/Core/src/Juvix/Core/IR/Evaluator.hs [hunk 4]
```diff
@@ -217,7 +217,7 @@ evalTerm g t = evalTermWith g rejectExts $ OnlyExts.onlyExtsT t
 -- becomes:
 --
 -- @
--- (ω | \x -> add x x : int -> int)
+-- (_ | \x -> add x x : int -> int)
```
Comment (mariari) [inline library/Core/src/Juvix/Core/IR/Evaluator.hs:220]: Again

File: library/Core/src/Juvix/Core/IR/Evaluator.hs [hunk 5]
```diff
@@ -289,7 +289,7 @@ toOnlyExtsE = extTransformE $ OnlyExts.injector
...
```
Comment (mariari) [inline library/Core/src/Juvix/Core/IR/Evaluator.hs:292]: Here as well

File: library/Core/test/HR/Pretty.hs [hunk 6]
```diff
@@ -58,43 +58,43 @@ bindTests =
...
       T.testCase "xAyBC' wide" $
         prettyAt 1000 xAyBC'
-          @?= "Π (ω | x : A) → Σ (ω | y : B) → C x y",
+          @?= "Π (_ | x : A) → Σ (_ | y : B) → C x y",
       T.testCase "xAyBC' narrow" $
         prettyAt 20 xAyBC'
-          @?= "Π (ω | x : A) →\n\
-              \Σ (ω | y : B) →\n\
+          @?= "Π (_ | x : A) →\n\
+              \Σ (_ | y : B) →\n\
```
Comment (mariari) [inline library/Core/test/HR/Pretty.hs:80]: We should find a symbol or say any, as it's unused does not imply any

File: library/Core/test/HR/Pretty.hs [hunk 7]
```diff
@@ -264,15 +264,15 @@ annTests =
     "Annotations"
     [ T.testCase "ann wide" $
         prettyAt 1000 ann
-          @?= "(ω | f x y : C x y : * 0)",
+          @?= "(_ | f x y : C x y : * 0)",
       T.testCase "ann narrow" $
         prettyAt 10 ann
-          @?= "(ω | f x y\n\
+          @?= "(_ | f x y\n\
               \   : C x y\n\
```
Comment (mariari) [inline library/Core/test/HR/Pretty.hs:271]: More spots to note

File: library/StandardLibrary/src/Juvix/Library/Usage.hs [hunk 8]
```diff
@@ -2,95 +2,124 @@
...
+  compare SAny SAny = EQ
 
 ------------------------------------------------------------------------------
 
 type instance PP.Ann Usage = ()
 
 instance PP.PrettySyntax Usage where
   pretty' (SNat π) = pure . PP.show $ π
-  pretty' Omega = pure "ω"
+  pretty' SAny = pure "_"
```
Comment (mariari) [inline library/StandardLibrary/src/Juvix/Library/Usage.hs:73]: Yeah here is where we have to chose

File: library/StandardLibrary/src/Juvix/Library/Usage.hs [hunk 9]
```diff
@@ -2,95 +2,124 @@
...
-pred (SNat x) = SNat (x - 1)
-pred Omega = Omega
+infixl 6 `minus`
 
 minus :: Usage -> Usage -> Maybe Usage
-minus (SNat i) (SNat j)
-  | i >= j = Just . SNat $ i - j
-minus Omega _ = Just Omega
+minus SAny _ = Just SAny
+minus (SNat π) (SNat ρ) | π >= ρ = Just . SNat $ π - ρ
```
Comment (mariari) [inline library/StandardLibrary/src/Juvix/Library/Usage.hs:109]: Probably best to leave `|` on the newline?

File: library/StandardLibrary/test/NameSymb.hs [hunk 10]
```diff
@@ -1,11 +1,15 @@
...
+--------------------------------------------------------------------------------
+
 import Juvix.Library
 import qualified Juvix.Library.NameSymbol as NameSymbol
 import qualified Test.Tasty as T
 import qualified Test.Tasty.HUnit as T
 import qualified Test.Tasty.QuickCheck as T
 
+--------------------------------------------------------------------------------
+
```
Comment (mariari) [inline library/StandardLibrary/test/NameSymb.hs:12]: Sections around imports aren't really needed as they have to come after the module with nothing in between due to how haskell handles modules

**Review by mariari — CHANGES_REQUESTED** [review]
Overall the PR looks good, there are a few changes I'd like to see back, but the biggest thing we should probably address is the use of `_` for denoting Any.

I'm not sure the `_` is appropriate here, as a `_` denotes unused rather than any/don't care.

File: library/StandardLibrary/src/Juvix/Library/Usage.hs (hunk 9 above)
Comment (jonaprieto) [reply library/StandardLibrary/src/Juvix/Library/Usage.hs:109]: OK.

File: library/StandardLibrary/test/NameSymb.hs (hunk 10 above)
Comment (jonaprieto) [reply library/StandardLibrary/test/NameSymb.hs:12]: This is a style issue. It's not a need, of course. So I can keep it as it was, no big deal. But, maybe,  the style guide we have should include more discussion on section, naming, comments, docs, etc. As the whole codebase is not consistent. But it's not a priority now, so let us keep it for later.

**Review by mariari — APPROVED** [review]
Ack

**Pull request merged**